File: arena/board.py
"""Results, and the two boards they add up to.

A result is written once, at the whistle, from the room's own log. Nothing a
client sends is stored as a score, and nothing here recomputes a match that has
already been scored -- a second whistle for the same room is a no-op, so a host
that reconnects and replays cannot pay somebody twice.

The two boards never merge. The README records the shipped squad as 0W-1D-7L,
so beating it and beating a person are not the same achievement.
"""

import json
import time

import rooms
import scoring

# What both boards are made of: a finished, ranked room. The workshop is not
# ranked, a match somebody ran fast is not ranked, and an abandoned match never
# reaches `finished`.
RANKED = "room.ranked = 1 AND room.status = 'finished'"


def record(conn, room):
    """Score a finished match and store one row per dugout. Returns the rows.

    Idempotent by the unique key on (room, player): if this room has already
    been scored, the stored rows come back untouched.
    """
    already = read(conn, room["id"])
    if already:
        return already

    facts = scoring.read(rooms.events(conn, room["id"]))
    seats = rooms.seated(conn, room["id"])
    ratings = _new_ratings(conn, room, seats, facts)

    now = time.time()
    for seat in seats:
        mine = facts[seat["team"]]
        scored = scoring.score_attack(mine)
        conn.execute(
            "INSERT INTO result (room_id, player_id, team, points, outcome,"
            "                    goals_for, goals_against, first_goal_ms, shouts,"
            "                    effective, rating, breakdown_json, computed_at) "
            "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            (room["id"], seat["player_id"], seat["team"], scored["points"],
             mine["outcome"], mine["goals_for"], mine["goals_against"],
             mine["first_goal_ms"], mine["shouts"], mine["effective"],
             ratings.get(seat["team"]), json.dumps(scored["breakdown"]), now),
        )
    conn.commit()
    return read(conn, room["id"])


def _new_ratings(conn, room, seats, facts):
    """Both managers' Elo after this match, or nothing if it was not a duel.

    A rating is a claim about one person against another, so the house side
    earns nobody one and a versus room that kicked off with an empty dugout
    earns nobody one either.
    """
    if room["mode"] != "versus" or len(seats) != len(scoring.TEAMS):
        return {}
    before = {seat["team"]: rating(conn, seat["player_id"]) for seat in seats}
    blue, red = scoring.TEAMS
    return {
        blue: scoring.rated(before[blue], before[red], facts[blue]["outcome"]),
        red: scoring.rated(before[red], before[blue], facts[red]["outcome"]),
    }


def rating(conn, player_id):
    """This player's Elo going into their next duel."""
    row = conn.execute(
        "SELECT rating FROM result WHERE player_id = ? AND rating IS NOT NULL "
        "ORDER BY computed_at DESC, id DESC LIMIT 1",
        (player_id,),
    ).fetchone()
    return row["rating"] if row else scoring.START_RATING


def read(conn, room_id):
    """One room's results, keyed by dugout. Empty until the whistle.

    This is the only read that carries the breakdown, because the results
    screen is the only place that shows one.
    """
    return {row["team"]: {**_run(row), "breakdown": json.loads(row["breakdown_json"])}
            for row in conn.execute(
                "SELECT r.*, p.display_name, p.email_masked, s.philosophy, room.code "
                "FROM result r "
                "JOIN room ON room.id = r.room_id "
                "JOIN player p ON p.id = r.player_id "
                "LEFT JOIN seat s ON s.room_id = r.room_id AND s.team = r.team "
                "WHERE r.room_id = ? ORDER BY r.team",
                (room_id,),
            )}


def solo(conn):
    """Score attack, best run per manager, highest first.

    Best run and not latest run: a manager who plays twice is trying to beat
    themselves, and a board that forgot their good one would punish them for
    having another go.
    """
    best = {}
    for row in conn.execute(
        "SELECT r.*, p.display_name, p.email_masked, s.philosophy, room.code "
        "FROM result r "
        "JOIN room ON room.id = r.room_id "
        "JOIN player p ON p.id = r.player_id "
        "LEFT JOIN seat s ON s.room_id = r.room_id AND s.team = r.team "
        f"WHERE {RANKED} AND room.mode = 'solo' "
        "ORDER BY r.points DESC, r.computed_at ASC, r.id ASC",
    ):
        # Ties go to whoever got there first, which is why the query orders on
        # the clock as well as the points.
        best.setdefault(row["player_id"], _run(row))
    return list(best.values())


def versus(conn):
    """Head to head, by wins and then goal difference.

    Not by rating. Elo is computed, stored and shown, but most people at an
    event play once or twice and one match is not a rating.
    """
    played = {}
    for row in conn.execute(
        "SELECT r.*, p.display_name, p.email_masked, room.code,"
        "       (SELECT p2.display_name FROM result r2"
        "          JOIN player p2 ON p2.id = r2.player_id"
        "         WHERE r2.room_id = r.room_id AND r2.player_id != r.player_id) AS against "
        "FROM result r "
        "JOIN room ON room.id = r.room_id "
        "JOIN player p ON p.id = r.player_id "
        f"WHERE {RANKED} AND room.mode = 'versus' "
        "ORDER BY r.computed_at ASC, r.id ASC",
    ):
        standing = played.setdefault(row["player_id"], _standing(row))
        standing["played"] += 1
        standing[row["outcome"]] += 1
        standing["goals_for"] += row["goals_for"]
        standing["goals_against"] += row["goals_against"]
        # The rows arrive oldest first, so the last one to land is the latest,
        # and the rating it carries is the one this manager holds now.
        standing["rating"] = row["rating"]
        standing["last"] = {"outcome": row["outcome"], "room": row["code"],
                            "goals_for": row["goals_for"],
                            "goals_against": row["goals_against"],
                            "against": row["against"]}

    table = list(played.values())
    for standing in table:
        standing["difference"] = standing["goals_for"] - standing["goals_against"]
    table.sort(key=lambda one: (-one["won"], -one["difference"], -one["goals_for"]))
    return table


def table_for(conn, mode):
    """Whichever board a match of this mode belongs on."""
    return solo(conn) if mode == "solo" else versus(conn)


def placing(conn, room, results):
    """Where this match leaves each dugout on the board it belongs to.

    Empty for a room that earns no place: the workshop, a match run fast, one
    still being played. The results screen still shows the breakdown for those,
    because what a practice run earned is worth reading; it just does not say
    where it ranks, because it ranks nowhere.
    """
    if not room["ranked"] or room["status"] != "finished":
        return {}
    table = table_for(conn, room["mode"])
    where = {row["player_id"]: index for index, row in enumerate(table)}
    standing = {}
    for team, result in results.items():
        index = where.get(result["player_id"])
        if index is None:
            continue
        standing[team] = {"rank": index + 1, "of": len(table)}
        if room["mode"] == "solo":
            # The board holds only a manager's best run, so this run is their
            # best exactly when the board is the one showing it.
            standing[team]["best"] = table[index]["room"] == room["code"]
    return standing


def top(conn, mode, most=3):
    """The head of the board this match belongs to, for the results screen."""
    return table_for(conn, mode)[:most]


def managers(conn):
    """How many people hold a ranked result at all, for the board's header."""
    return conn.execute(
        "SELECT COUNT(DISTINCT r.player_id) AS them FROM result r "
        f"JOIN room ON room.id = r.room_id WHERE {RANKED}"
    ).fetchone()["them"]


def _run(row):
    """One stored result, as every reader but the results screen wants it."""
    return {
        "player_id": row["player_id"],
        "name": row["display_name"],
        "email": row["email_masked"],
        "team": row["team"],
        "philosophy": row["philosophy"],
        "room": row["code"],
        "points": row["points"],
        "outcome": row["outcome"],
        "goals_for": row["goals_for"],
        "goals_against": row["goals_against"],
        "first_goal_ms": row["first_goal_ms"],
        "shouts": row["shouts"],
        "effective": row["effective"],
        "rating": row["rating"],
    }


def _standing(row):
    return {"player_id": row["player_id"], "name": row["display_name"],
            "email": row["email_masked"], "played": 0,
            "won": 0, "drew": 0, "lost": 0,
            "goals_for": 0, "goals_against": 0, "difference": 0,
            "rating": row["rating"], "last": None}

File: arena/static/board.js
/**
 * The standings, on a wall or in a hand.
 *
 * Two boards that never merge, because beating the shipped squad and beating
 * a person are not the same achievement. Nothing here computes a point: the
 * arena scored every match from its own log, and this page renders what it
 * says.
 *
 * It spends most of an event unattended, so it swaps boards on its own and
 * refetches whenever a match ends anywhere in the venue. A click stops the
 * cycle, because somebody reading a table should not have it slide away.
 */

import { get } from "/static/api.js";
import { icon } from "/static/dom.js";
import { figure } from "/static/words.js";
import { openWall } from "/static/socket.js";

const CYCLE_MS = 12_000;
const MEDALS = ["🥇", "🥈", "🥉"];
const PODIUM = ["g", "s", "br"];   // gold in the middle, which is what `order` does
const OUTCOME = { won: "Won", drew: "Drew", lost: "Lost" };
const OUTCOME_TAG = { won: "w", drew: "d", lost: "l" };

const el = (id) => document.getElementById(id);
const tabs = { solo: el("tab-solo"), versus: el("tab-versus") };

let showing = "solo";
let pinned = false;      // somebody clicked a tab, so the page stopped cycling
let cycle = null;

// On a wall or in a hand, decided by the one thing that cannot be got wrong by
// whoever linked here. The big screen frames this page under its lobby, and a
// wall has nowhere to go home to; a phone arrived from home, the join form or
// the result sheet, and every one of those was a one-way trip.
if (window.self === window.top) el("board-home").hidden = false;

for (const [which, tab] of Object.entries(tabs)) {
  tab.addEventListener("click", () => {
    pinned = true;
    el("tick").hidden = true;
    clearTimeout(cycle);
    show(which);
  });
}

/**
 * The big screen framing this page, telling it which board to hold up.
 *
 * Only ever from the page that framed it, and only ever the one message. The
 * announcer talks over this frame for forty seconds, and the twelve-second
 * cycle would otherwise slide the board away in the middle of a sentence
 * about it.
 */
window.addEventListener("message", (note) => {
  if (note.origin !== location.origin) return;
  if (!note.data || note.data.type !== "board.show") return;
  if (note.data.pinned === false) return unpin();
  pinned = true;
  el("tick").hidden = true;
  clearTimeout(cycle);
  show(note.data.board === "versus" ? "versus" : "solo");
});

function unpin() {
  pinned = false;
  el("tick").hidden = false;
  turn();
}

load();
listen();

function listen() {
  // The wall topic carries a message at every kick-off and every whistle, which
  // is exactly when a board changes. Polling would either lag the venue or
  // hammer it, and the socket is already there.
  openWall({
    onMessage: (message) => { if (message.type === "wall") load(); },
    onDrop: (reason, permanent) => { if (permanent) complain(reason); },
  });
}

async function load() {
  try {
    const board = await get("/api/board");
    el("problem").hidden = true;
    el("where").textContent =
      `arena · board · ${count(board.managers, "manager", "managers", "nobody yet")}`;
    drawSolo(board.solo);
    drawVersus(board.versus);
    if (!pinned) turn();
  } catch (failure) {
    complain(failure.message);
  }
}

/* ── Which board is up ──────────────────────────────────────────────── */

function show(which) {
  showing = which;
  for (const [name, tab] of Object.entries(tabs)) {
    tab.setAttribute("aria-selected", String(name === which));
    el(`view-${name}`).hidden = name !== which;
  }
}

function turn() {
  clearTimeout(cycle);
  restartTick();
  cycle = setTimeout(() => {
    show(showing === "solo" ? "versus" : "solo");
    turn();
  }, CYCLE_MS);
}

function restartTick() {
  const bar = el("tick").firstElementChild;
  // Restarted by hand rather than by CSS class, because a fetch landing
  // mid-cycle resets the clock and the bar has to agree with it.
  bar.style.animation = "none";
  void bar.offsetWidth;
  bar.style.animation = `tick ${CYCLE_MS}ms linear forwards`;
}

/* ── Score attack ───────────────────────────────────────────────────── */

function drawSolo(runs) {
  podium("podium-solo", runs, (run) => ({
    lines: [`${count(run.goals_for, "goal", "goals", "No goals")} · ${run.goals_against} conceded`,
            run.first_goal_ms === null ? "never scored" : `first at ${clock(run.first_goal_ms)}`],
    score: figure(run.points),
    unit: "points",
  }));

  table("table-solo", runs, {
    empty: ["Nobody has run at the house side yet",
            "Scan the code on the arena screen and the first place is yours."],
    columns: [
      { head: "Opening shape", draw: (run) => cell("", titled(run.philosophy)) },
      { head: "Goals", draw: (run) => cell("num", `${run.goals_for}-${run.goals_against}`) },
      { head: "First goal",
        draw: (run) => cell("num", run.first_goal_ms === null ? "-" : clock(run.first_goal_ms)) },
      // "0 of 0" is a column of noise on a board where most runs are silent.
      { head: "Shouts that scored",
        draw: (run) => cell("num", run.shouts ? `${run.effective} of ${run.shouts}` : "-") },
    ],
    total: { head: "Points", draw: (run) => cell("tot", figure(run.points)) },
  });
}

/* ── Head to head ───────────────────────────────────────────────────── */

function drawVersus(table_) {
  podium("podium-versus", table_, (standing) => ({
    lines: [`${count(standing.played, "match", "matches", "none played")} · ${signed(standing.difference)} goal difference`],
    score: `${standing.won} - ${standing.drew} - ${standing.lost}`,
    unit: "won drew lost",
  }));

  table("table-versus", table_, {
    empty: ["No duels yet",
            "Two phones in one room and the winner takes this board."],
    columns: [
      { head: "Played", draw: (one) => cell("num", String(one.played)) },
      { head: "Record", draw: (one) => cell("num", `${one.won}-${one.drew}-${one.lost}`) },
      { head: "Goals", draw: (one) => cell("num", `${one.goals_for}-${one.goals_against}`) },
      { head: "Difference", draw: (one) => cell("num", signed(one.difference)) },
      { head: "Last match", draw: (one) => lastMatch(one.last) },
    ],
    // A duel one dugout never turned up for pays no rating, so there can be a
    // standing here with nothing to show in this column.
    total: { head: "Rating",
             draw: (one) => cell("tot", one.rating === null ? "-"
                                                            : figure(Math.round(one.rating))) },
  });
}

function lastMatch(last) {
  const box = document.createElement("td");
  if (!last) return box;
  const tag = document.createElement("span");
  tag.className = `tag ${OUTCOME_TAG[last.outcome]}`;
  tag.textContent = OUTCOME[last.outcome];
  box.append(tag, ` ${last.goals_for}-${last.goals_against}`);
  // A duel always has somebody on the other side, but a name can be missing if
  // the other dugout's result has not been written; saying "v " and stopping
  // would read as a bug.
  if (last.against) box.append(` v ${last.against}`);
  return box;
}

/* ── The two shapes both boards are drawn in ────────────────────────── */

function podium(into, rows, describe) {
  const three = rows.slice(0, 3);
  el(into).replaceChildren(...three.map((row, place) => {
    const pod = document.createElement("div");
    pod.className = `pod ${PODIUM[place]}`;
    const medal = icon(MEDALS[place]);
    medal.classList.add("medal");
    const name = document.createElement("div");
    name.className = "nm";
    name.textContent = row.name;

    const told = describe(row);
    const state = document.createElement("div");
    state.className = "st";
    told.lines.forEach((line, index) => {
      if (index) state.append(document.createElement("br"));
      state.append(line);
    });
    const score = document.createElement("div");
    score.className = "sc";
    score.append(told.score,
                 Object.assign(document.createElement("small"), { textContent: told.unit }));

    pod.append(medal, name, ...masked("div", "em", row.email), state, score);
    return pod;
  }));
}

/**
 * One board as a table: rank, manager, the detail, and the number it is on.
 *
 * The detail columns are marked `.det` and a phone drops them, because eight
 * columns on a 375px screen either shrink to nothing or scroll sideways under
 * a page that already scrolls down. Rank, who, and the figure the board is
 * ordered by survive, which is the board.
 */
function table(into, rows, { empty, columns, total }) {
  if (!rows.length) return el(into).replaceChildren(nobody(...empty));

  const header = document.createElement("tr");
  header.append(head("#", "rk"), head("Manager"),
                ...columns.map((column) => head(column.head, "det")),
                head(total.head, "tot"));

  const body = document.createElement("tbody");
  body.append(...rows.map((one, index) => {
    const line = document.createElement("tr");
    line.append(cell(index ? "rk" : "rk top", String(index + 1)), who(one));
    for (const column of columns) line.append(detail(column.draw(one)));
    line.append(total.draw(one));
    return line;
  }));

  const grid = document.createElement("table");
  grid.className = "table";
  grid.append(document.createElement("thead"), body);
  grid.firstElementChild.append(header);
  el(into).replaceChildren(grid);
}

function head(label, className) {
  const th = document.createElement("th");
  if (className) th.className = className;
  th.textContent = label;
  return th;
}

function detail(box) {
  box.classList.add("det");
  return box;
}

function who(one) {
  const box = cell("who-cell", "");
  const name = document.createElement("b");
  name.textContent = one.name;
  box.append(name, ...masked("span", "", one.email));
  return box;
}

/**
 * The masked address under a manager's name, as nothing or as one element.
 *
 * An address is optional, so a great many rows have none, and an empty element
 * is not the same as no element: on the podium it is a margin and a line box
 * where the managers either side have a line of text, which reads as a
 * rendering fault rather than as somebody who kept their address to themselves.
 */
function masked(tag, className, address) {
  if (!address) return [];
  const line = document.createElement(tag);
  if (className) line.className = className;
  line.textContent = address;
  return [line];
}

function cell(className, text) {
  const box = document.createElement("td");
  if (className) box.className = className;
  box.textContent = text;
  return box;
}

function nobody(title, line) {
  const box = document.createElement("div");
  box.className = "board-empty";
  box.append(Object.assign(document.createElement("b"), { textContent: title }),
             Object.assign(document.createElement("span"), { textContent: line }));
  return box;
}

/* ── Numbers as a manager reads them ────────────────────────────────── */

const signed = (value) => (value > 0 ? `+${value}` : String(value));
const titled = (word) => (word ? word.charAt(0).toUpperCase() + word.slice(1) : "-");

// Both forms spelled out rather than an -s rule, which turns two matches into
// two matchs.
function count(many, one, more, none) {
  return many ? `${many} ${many === 1 ? one : more}` : none;
}

function clock(ms) {
  const whole = Math.max(0, Math.round(ms / 1000));
  return `${Math.floor(whole / 60)}:${String(whole % 60).padStart(2, "0")}`;
}

function complain(message) {
  el("problem").textContent = message;
  el("problem").hidden = false;
}
